src/dso/dso_transport.cpp

```diff
@@ -0,0 +1,552 @@
+        {
+            readLen = mbedtls_net_recv(&mCtx, buf, std::min(sizeof(buf), mNeedBytes));
+
+            if (readLen <= 0)
+            {
+                HandleMbedTlsError(readLen);
+                VerifyOrExit(readLen != 0);
+                VerifyOrExit(readLen != MBEDTLS_ERR_SSL_WANT_READ && readLen != MBEDTLS_ERR_SSL_WANT_WRITE);
+                otbrLogWarning("Failed to receive message: %d", readLen);
+                ExitNow();
+            }
+            SuccessOrExit(otMessageAppend(mPendingMessage, buf, readLen));
+            mNeedBytes -= readLen;
+
+            if (!mNeedBytes)
+            {
+                otPlatDsoHandleReceive(mConnection, mPendingMessage);
+                mPendingMessage = nullptr;
+            }
+        }
+        else
+        {
+            assert(mLengthBuffer.size() < sizeof(uint16_t));
+            assert(mPendingMessage == nullptr);
+
+            readLen = mbedtls_net_recv(&mCtx, buf, sizeof(uint16_t) - mLengthBuffer.size());
+
+            if (readLen <= 0)
+            {
+                HandleMbedTlsError(readLen);
+                VerifyOrExit(readLen != 0);
+                VerifyOrExit(readLen != MBEDTLS_ERR_SSL_WANT_READ && readLen != MBEDTLS_ERR_SSL_WANT_WRITE);
+                otbrLogWarning("Failed to receive message: %d", readLen);
+                ExitNow();
+            }
+            for (int i = 0; i < readLen; ++i)
+            {
+                mLengthBuffer.push_back(buf[i]);
+            }
+            if (mLengthBuffer.size() == sizeof(uint16_t))
+            {
+                mNeedBytes      = mLengthBuffer[0] << 8 | mLengthBuffer[1];
+                mPendingMessage = otIp6NewMessage(otPlatDsoGetInstance(mConnection), nullptr);
+                mLengthBuffer.clear();
+                VerifyOrExit(mPendingMessage != nullptr);
+            }
+        }
+    }
+
+exit:
+    return;
+}
+
+void DsoAgent::DsoConnection::HandleMbedTlsError(int aError)
+{
+    VerifyOrExit(aError < 0);
+
+    switch (aError)
+    {
+    case MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY:
+        otPlatDsoHandleDisconnected(mConnection, OT_PLAT_DSO_DISCONNECT_MODE_GRACEFULLY_CLOSE);
+        mState = State::kDisabled;
+        break;
+    case MBEDTLS_ERR_NET_CONN_RESET:
+        otPlatDsoHandleDisconnected(mConnection, OT_PLAT_DSO_DISCONNECT_MODE_FORCIBLY_ABORT);
+        mState = State::kDisabled;
+        break;
+    case MBEDTLS_ERR_SSL_WANT_READ:
+    case MBEDTLS_ERR_SSL_WANT_WRITE:
+        break;
+    default:
+        otPlatDsoHandleDisconnected(mConnection, OT_PLAT_DSO_DISCONNECT_MODE_FORCIBLY_ABORT);
+        mState = State::kDisabled;
+        break;
+    }
+
+exit:
+    return;
+}
+
+void DsoAgent::DsoConnection::UpdateStateBySocketState(void)
+{
+    int       optVal;
+    socklen_t optLen;
+
+    otbrLogDebug("Updating state");
+
+    if (!getsockopt(GetFd(), SOL_SOCKET, SO_ERROR, &optVal, &optLen))
+    {
+        if (!optVal)
+        {
+            mState = State::kConnected;
```

Comment:
Please **log** state changes similar to https://github.com/openthread/openthread/blob/main/src/core/net/srp_client.cpp#L71